File: packages/simplex-chat-nodejs/src/bot.ts
import {T} from "@simplex-chat/types"
import * as api from "./api"
import * as core from "./core"
import * as util from "./util"
import equal = require("fast-deep-equal")

export interface BotDbOpts {
  dbFilePrefix: string // two schema files will be named <prefix>_chat.db and <prefix>_agent.db
  dbKey?: string
  confirmMigrations?: core.MigrationConfirmation
}

export interface BotOptions {
  createAddress?: boolean
  updateAddress?: boolean
  updateProfile?: boolean
  addressSettings?: api.BotAddressSettings
  allowFiles?: boolean
  commands?: T.ChatBotCommand[] // commands to show in client UI
  useBotProfile?: boolean // create profile not marked as a bot, with default preferences  
  logContacts?: boolean
  logNetwork?: boolean
}

const defaultOpts: Required<BotOptions> = {
  createAddress: true,
  updateAddress: true,
  updateProfile: true,
  addressSettings: api.defaultBotAddressSettings,
  allowFiles: false,
  commands: [],
  useBotProfile: true,
  logContacts: true,
  logNetwork: false
}

export interface BotConfig {
  profile: T.Profile,
  dbOpts: BotDbOpts,
  options: BotOptions,
  onMessage?: (chatItem: T.AChatItem, content: T.MsgContent) => void | Promise<void>,
  // command handlers can be different from commands to be shown in client UI
  onCommands?: {[K in string]?: ((chatItem: T.AChatItem, command: util.BotCommand) => void | Promise<void>)},
  // If you use `onMessage` and to subscribe "newChatItems" event, exclude content messages from processing
  // If you use `onCommands` and to subscribe "newChatItems" event, exclude commands from processing
  events?: api.EventSubscribers
}

export async function run({profile, dbOpts, options = defaultOpts, onMessage, onCommands = {}, events = {}}: BotConfig): Promise<[api.ChatApi, T.User]> {
  const bot = await api.ChatApi.init(dbOpts.dbFilePrefix, dbOpts.dbKey || "", dbOpts.confirmMigrations || core.MigrationConfirmation.YesUp)
  const opts = fullOptions(options)
  if (onMessage) subscribeMessages(bot, onMessage)
  if (Object.keys(onCommands).length > 0) subscribeCommands(bot, onCommands)
  if (Object.keys(events).length > 0) bot.on(events)
  subscribeLogEvents(bot, opts)
  const botProfile = mkBotProfile(profile, opts)
  const user = await createBotUser(bot, botProfile)
  await bot.startChat()
  const address = await createOrUpdateAddress(bot, user, opts)  
  const addressLink = util.contactAddressStr(address.connLinkContact)
  console.log(`Bot address: ${addressLink}`)
  if (opts.useBotProfile) botProfile.contactLink = addressLink
  await updateBotUserProfile(bot, user, botProfile, opts)  
  return [bot, user]
}

function fullOptions(options: BotOptions): Required<BotOptions> {
  const opts =  {
    createAddress: options.createAddress ?? defaultOpts.createAddress,
    updateAddress: options.updateAddress ?? defaultOpts.updateAddress,
    updateProfile: options.updateProfile ?? defaultOpts.updateProfile,
    addressSettings: options.addressSettings ?? defaultOpts.addressSettings,
    allowFiles: options.allowFiles ?? defaultOpts.allowFiles,
    commands: options.commands ?? defaultOpts.commands,
    useBotProfile: options.useBotProfile ?? defaultOpts.useBotProfile,
    logContacts: options.logContacts ?? defaultOpts.logContacts,
    logNetwork: options.logNetwork ?? defaultOpts.logNetwork
  }
  const welcomeMessage = opts.addressSettings.welcomeMessage ?? defaultOpts.addressSettings.welcomeMessage
  opts.addressSettings = {
    autoAccept: opts.addressSettings.autoAccept ?? defaultOpts.addressSettings.autoAccept,
    welcomeMessage: typeof welcomeMessage === "string" ? {type: "text", text: welcomeMessage} : welcomeMessage,
    businessAddress: opts.addressSettings.businessAddress ?? defaultOpts.addressSettings.businessAddress
  }
  return opts
}

function mkBotProfile(profile: T.Profile, opts: Required<BotOptions>): T.Profile {
  if (opts.useBotProfile) {
    const prefs = profile.preferences || {}
    if (prefs.files || prefs.calls || prefs.voice || prefs.commands) {
      console.log("Option useBotProfile is enabled and profile preferences used for files, calls, voice or commands, exiting")
      process.exit()
    }
    prefs.files = {allow: opts.allowFiles ? T.FeatureAllowed.Yes : T.FeatureAllowed.No}
    prefs.calls = {allow: T.FeatureAllowed.No}
    prefs.voice = {allow: T.FeatureAllowed.No}
    prefs.commands = opts.commands
    profile.preferences = prefs
    profile.peerType = T.ChatPeerType.Bot
  } else if (opts.commands.length > 0) {
      console.log("Option useBotProfile is disabled and commands are passed, exiting")
      process.exit()
  }
  return profile 
}

function subscribeMessages(bot: api.ChatApi, onMessage: (chatItem: T.AChatItem, content: T.MsgContent) => void | Promise<void>) {
  bot.on("newChatItems", async ({chatItems}) => {
    for (const ci of chatItems) {
      if (ci.chatItem.content.type === "rcvMsgContent") {
        try {
          const p = onMessage(ci, ci.chatItem.content.msgContent)
          if (p instanceof Promise) await p
        } catch (e) {
          console.log(`message processing error`, e)
        }
      }
    }
  })
}

function subscribeCommands(bot: api.ChatApi, commands: {[K in string]?: ((chatItem: T.AChatItem, command: util.BotCommand) => void | Promise<void>)}) {
  bot.on("newChatItems", async (evt) => {
    for (const ci of evt.chatItems) {
      const cmd = util.ciBotCommand(ci.chatItem)
      if (cmd) {
        const cmdFunc = commands[cmd.keyword] || commands[""]
        if (cmdFunc) {
          try {
            const p = cmdFunc(ci, cmd)
            if (p instanceof Promise) await p
          } catch(e) {
            console.log(`${cmd} command processing error`, e)
          }
        }
      }
    }
  })    
}

function subscribeLogEvents(bot: api.ChatApi, opts: Required<BotOptions>) {
  if (opts.logContacts) {
    bot.on({
      "contactConnected": ({contact}) => console.log(`${contact.profile.displayName} connected`),
      "contactDeletedByContact": ({contact}) => console.log(`${contact.profile.displayName} deleted connection with bot`)
    })      
  }
  if (opts.logNetwork) {
    bot.on({
      "hostConnected": ({transportHost}) => console.log(`connected server ${transportHost}`),
      "hostDisconnected": ({transportHost}) => console.log(`diconnected server ${transportHost}`),
      "subscriptionStatus": ({subscriptionStatus, connections}) => console.log(`${connections.length} subscription(s) ${subscriptionStatus.type}`)
    })          
  }
}

async function createBotUser(bot: api.ChatApi, profile: T.Profile): Promise<T.User> {
  let user = await bot.apiGetActiveUser()
  if (!user) {
    console.log("No active user in dabase, creating...")
    user = await bot.apiCreateActiveUser(profile)
  }
  console.log("Bot user: ", user.profile.displayName)
  return user    
}
    
async function createOrUpdateAddress(bot: api.ChatApi, user: T.User, opts: Required<BotOptions>): Promise<T.UserContactLink> {
  const {userId} = user
  let address = await bot.apiGetUserAddress(userId)
  if (!address) {
    if (opts.createAddress) {
      console.log("Bot has no address, creating...")
      await bot.apiCreateUserAddress(userId)
      address = await bot.apiGetUserAddress(userId)
      if (!address) {
        console.log("Failed reading created user address, exiting")
        process.exit()
      }
    } else {
      console.log("Bot has no address, exiting")
      process.exit()
    }
  }
  
  const addressSettings = opts.addressSettings || defaultOpts.addressSettings
  if (!equal(util.botAddressSettings(address), addressSettings)) {
    console.log(util.botAddressSettings(address), addressSettings)
    if (opts.updateAddress) {
      console.log("Bot address settings changed, updating...")
      await bot.apiSetAddressSettings(userId, addressSettings)
    } else {
      console.log("Bot address settings changed")
    }
  }
  
  return address    
}

async function updateBotUserProfile(bot: api.ChatApi, user: T.User, profile: T.Profile, opts: Required<BotOptions>): Promise<void> {
  const {userId} = user
  if (!equal(util.fromLocalProfile(user.profile), profile)) {
    if (opts.updateProfile) {
      console.log("Bot profile changed, updating...")
      const summary = await bot.apiUpdateProfile(userId, profile)
      console.log(
        summary
        ? `Bot profile updated: ${summary.updateSuccesses} updated contact(s), ${summary.updateFailures} failed contact update(s).`
        : "Unexpected: profile did not change!"
      )
    } else {
      console.log("Bot profile changed")
    }
  }  
}

File: packages/simplex-chat-python/src/simplex_chat/bot.py
"""`Bot` — Client extended with server-side features (address, auto-accept, commands)."""

from __future__ import annotations

from dataclasses import dataclass

from . import util
from .api import Db
from .client import (
    BotProfile,
    ChatMessage,
    Client,
    CommandHandler,
    EventHandler,
    FileMessage,
    ImageMessage,
    LinkMessage,
    Message,
    MessageHandler,
    Middleware,
    ParsedCommand,
    Profile,
    ReportMessage,
    TextMessage,
    UnknownMessage,
    VideoMessage,
    VoiceMessage,
    log,
)
from .core import MigrationConfirmation
from .types import T


@dataclass(slots=True)
class BotCommand:
    keyword: str
    label: str
    params: str = ""


class Bot(Client):
    """SimpleX bot — Client extended with server-side features.

    On top of `Client` (identity + messaging + connect_to/send_and_wait/events),
    a Bot:
      - creates and announces its own contact address
      - auto-accepts incoming contact requests (configurable)
      - advertises a list of slash-commands in its profile preferences
      - sets `peerType=bot` and disables calls/voice in profile prefs
      - sends a `welcome` message to new contacts via the auto-reply address setting

    If you want just identity + messaging without any of that, use `Client`
    directly.
    """

    def __init__(
        self,
        *,
        profile: Profile,
        db: Db,
        welcome: str | T.MsgContent | None = None,
        commands: list[BotCommand] | None = None,
        confirm_migrations: MigrationConfirmation = MigrationConfirmation.YES_UP,
        create_address: bool = True,
        update_address: bool = True,
        update_profile: bool = True,
        auto_accept: bool = True,
        business_address: bool = False,
        allow_files: bool = False,
        log_contacts: bool = True,
        log_network: bool = False,
    ) -> None:
        super().__init__(
            profile=profile,
            db=db,
            confirm_migrations=confirm_migrations,
            update_profile=update_profile,
            log_contacts=log_contacts,
            log_network=log_network,
        )
        self._welcome = welcome
        self._commands = commands or []
        self._create_address = create_address
        self._update_address = update_address
        self._auto_accept = auto_accept
        self._business_address = business_address
        self._allow_files = allow_files

    # ------------------------------------------------------------------ #
    # Profile + address sync (overrides hooks in Client)
    # ------------------------------------------------------------------ #

    async def _post_start(self, user: T.User) -> None:
        """Bots sync address first, then embed the link in the profile."""
        link = await self._sync_address(user)
        await self._maybe_sync_profile(user, contact_link=link)

    async def _sync_address(self, user: T.User) -> str | None:
        """Address sync. Returns the public link if any, for embedding in the profile."""
        api = self.api
        user_id = user["userId"]

        address = await api.api_get_user_address(user_id)
        if address is None:
            if self._create_address:
                log.info("Bot has no address, creating...")
                await api.api_create_user_address(user_id)
                address = await api.api_get_user_address(user_id)
                if address is None:
                    raise RuntimeError("Failed reading newly created user address")
            else:
                log.warning("Bot has no address")

        link: str | None = None
        if address is not None:
            link = util.contact_address_str(address["connLinkContact"])
            log.info("Bot address: %s", link)

        # Address settings (auto-accept + welcome message). Mirrors bot.ts:185-194.
        # autoAccept present → accept; absent → no auto-accept (mirrors Node bot.ts).
        if address is not None and self._update_address:
            desired: T.AddressSettings = {"businessAddress": self._business_address}
            if self._auto_accept:
                desired["autoAccept"] = {"acceptIncognito": False}
            if self._welcome is not None:
                desired["autoReply"] = (
                    {"type": "text", "text": self._welcome}
                    if isinstance(self._welcome, str)
                    else self._welcome
                )
            if address["addressSettings"] != desired:
                log.info("Bot address settings changed, updating...")
                await api.api_set_address_settings(user_id, desired)

        return link

    def _profile_to_wire(self) -> T.Profile:
        """Bot profile: base profile + peerType=bot, command list, calls/voice prefs disabled.

        Mirrors Node `mkBotProfile` (bot.ts:88-102).
        """
        p = super()._profile_to_wire()
        prefs: T.Preferences = {
            "calls": {"allow": "no"},
            "voice": {"allow": "no"},
            "files": {"allow": "yes" if self._allow_files else "no"},
        }
        if self._commands:
            prefs["commands"] = [
                {"type": "command", "keyword": c.keyword, "label": c.label, "params": c.params}
                if c.params else
                {"type": "command", "keyword": c.keyword, "label": c.label}
                for c in self._commands
            ]
        p["preferences"] = prefs
        p["peerType"] = "bot"
        return p


__all__ = [
    "Bot",
    "BotCommand",
    "BotProfile",
    "ChatMessage",
    "Client",
    "CommandHandler",
    "EventHandler",
    "FileMessage",
    "ImageMessage",
    "LinkMessage",
    "Message",
    "MessageHandler",
    "Middleware",
    "ParsedCommand",
    "Profile",
    "ReportMessage",
    "TextMessage",
    "UnknownMessage",
    "VideoMessage",
    "VoiceMessage",
]
